lib/internal/modules/esm/loader.js

```diff
@@ -607,6 +550,36 @@ class ESMLoader {
       format,
       source,
     } = loaded;
+    let responseURL = loaded.responseURL;
+
+    if (responseURL === undefined) {
+      responseURL = url;
+    }
+
+    if (typeof responseURL !== 'string') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'undefined or a string',
+        hookErrIdentifier,
+        'responseURL',
+        responseURL,
+      );
+    }
+
+    let responseURLObj;
+    try {
+      responseURLObj = new URL(responseURL);
+    } catch {
+      // Continue regardless of error.
+    }
+
+    if (responseURLObj?.href !== responseURL) {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'a fully resolved URL string',
+        hookErrIdentifier,
+        'responseURL',
+        responseURL,
+      );
+    }
```

Comment:
Yes because you could return a non-canonical URL so it's trying to explicitly catch that.

In addition, the error thrown by `new URL` is awful and gives no context at all, in general we always want to wrap that error and never send it out directly where possible.